### PR TITLE
Migrated the Edraak/ForUs app from dogwood with tests and quality fixes

### PR DESCRIPTION
**Pull request by Salomari1987**
### Description

TASK: [Migrate edraak_forus module, with a feature flag](https://app.asana.com/0/198215864990333/276104526908807)

Migrated Edraak_forus

### CODE REVIEW

File: common/djangoapps/course_modes/helpers.py [hunk 1]
```diff
@@ -13,6 +15,43 @@
...
+
+    url = '{root}{lang}{page}'.format(
+        root=settings.MKTG_URLS['ROOT'],
+        lang=lang,
+        page=page
+    )
+    return url
+
+
+def get_mktg_for_course(page, course_id):
```
Comment (OmarIthawi) [inline common/djangoapps/course_modes/helpers.py]: Not sure if those are required anymore. edX might have added alternative solutions.

Will check them later.

File: common/djangoapps/student/helpers.py [hunk 2]
```diff
@@ -198,6 +220,75 @@ def auth_pipeline_urls(auth_entry, redirect_url=None):
...
+def _update_email_opt_in(request, org):
+    """Helper function used to hit the profile API if email opt-in is enabled."""
+
+    email_opt_in = request.POST.get('email_opt_in')
+    if email_opt_in is not None:
+        email_opt_in_boolean = email_opt_in == 'true'
+        preferences_api.update_email_opt_in(request.user, org, email_opt_in_boolean)
+
+
+def enroll(user, course_id, request, check_access):
```
Comment (OmarIthawi) [inline common/djangoapps/student/helpers.py]: @Salomari1987 This refactoring is old. It has been done in 2014, when there was mostly no API in the platform. Please take a look at the [enrollment api](http://edx.readthedocs.io/projects/edx-platform-api/en/latest/enrollment/overview.html).

Take a look at the way I've used the [`update_account_settings`](https://github.com/Edraak/edraak-platform/pull/40/files#diff-8f391fa7e6ff5b7e12c477ca0153f52f) api internally (without having to do a seperate HTTP request).

File: common/djangoapps/student/migrations/0007_auto_20170227_0530.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
```
Comment (OmarIthawi) [inline common/djangoapps/student/migrations/0007_auto_20170227_0530.py]: ~~Perfect auto migration :rofl:~~

I take it back. Don't add migration, it will break future updates by edX in a very horrible way that needs a migration `--merge`. Just keep it at the application level.


**Protip:** Most probably the `choices` parameter is being handled at the Django ORM level, not at the database level. So not sure if it is required to have this migration. We can ignore the migration warning for this specific case to make sure we don't reck our data in future migrations.

Comment (OmarIthawi) [reply common/djangoapps/student/migrations/0007_auto_20170227_0530.py]: As you've suggested, we could handle our gender choices at the application level and keep the models unmodified.

File: common/lib/xmodule/xmodule/course_module.py [hunk 4]
```diff
@@ -1018,6 +1019,39 @@ def has_ended(self):
         """
         return course_metadata_utils.has_course_ended(self.end)
 
+    def enrollment_has_ended(self):
```
Comment (OmarIthawi) [inline common/lib/xmodule/xmodule/course_module.py]: Since we need this in a single module, I suggest to move this to `edraak_forus.helpers` to avoid future migration issues whenever possible.

File: lms/djangoapps/edraak_forus/admin.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        """
+        Returns link to Edraak profile
+        """
+        try:
+            edraak_profile = UserProfile.objects.get(user=profile.user)
+        except UserProfile.DoesNotExist:
+            return None
+
+        return HTML(u'<a href="{url}">Edraak Profile</a>').format(
+            url=reverse('admin:student_userprofile_change', args=[edraak_profile.pk]),
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/admin.py]: You might want to use `Text()` for the format params like `url`, this is especially important because the linter won't check for this outside `*.html` files.

File: lms/djangoapps/edraak_forus/helpers.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+from student.models import UserProfile
+
+
+DATE_TIME_FORMAT = '%Y-%m-%dT%H:%M:%S'
+
+
+log = logging.getLogger(__name__)
+
+# pylint: disable=invalid-name
+ordered_hmac_keys = (
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/helpers.py]: Just make it all caps. Avoid all pylint pragmas whenever possible.

File: lms/djangoapps/edraak_forus/helpers.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    'country',
+    'level_of_education',
+    'gender',
+    'year_of_birth',
+    'lang',
+    'time',
+)
+
+
+def is_enabled_language(lang):
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/helpers.py]: 🤔 not sure if this the best way. Please check how the code around `SHOW_LANGUAGE_SELECTOR` does similar checks.

File: lms/djangoapps/edraak_forus/helpers.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+def calculate_hmac(msg_to_hash):
+    """
+    Returns hmac hexdigest of messages
+    """
+    secret_key = settings.FORUS_AUTH_SECRET_KEY
+    dig = hmac.new(secret_key.encode('utf-8'), msg_to_hash.encode('utf-8'), digestmod=sha256)
+    return dig.hexdigest()
+
+
+class ValidateForusParamsValues(object):
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/helpers.py]: Now it's a class. Rename to `ForusValidator`, or any other noun.

File: lms/djangoapps/edraak_forus/helpers.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+    """
+    Validates ForUs params values
+    """
+    def __init__(self, params):
+        self.params = params
+        self.errors = defaultdict(lambda: [])
+
+    # pylint: disable=missing-docstring
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/helpers.py]: `' '.join(['Please'] * 10)` do what pylint wants.

File: lms/djangoapps/edraak_forus/helpers.py [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+    secret_key = settings.FORUS_AUTH_SECRET_KEY
+    dig = hmac.new(secret_key.encode('utf-8'), msg_to_hash.encode('utf-8'), digestmod=sha256)
+    return dig.hexdigest()
+
+
+class ValidateForusParamsValues(object):
+    """
+    Validates ForUs params values
+    """
+    def __init__(self, params):
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/helpers.py]: Typical enterprise pile of code :rofl:

File: lms/djangoapps/edraak_forus/helpers.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+        self.validate_user_email()
+        self.validate_user_profile()
+        self.validate_course()
+        self.validate_request_time()
+
+        if len(self.errors):
+            raise ValidationError(self.errors)
+
+
+class ValidateForusParams(object):
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/helpers.py]: I think this needs refactoring with the other classes.

File: lms/djangoapps/edraak_forus/models.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
+"""
+Edraak ForUs models
+"""
+import logging
+from django.db import models
+from django.contrib.auth.models import User
+from django.utils.translation import ugettext_lazy as _
+from django.utils import timezone
+
+log = logging.getLogger(__name__)
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/models.py]: ~~It's called `logger` by convention.~~

File: lms/djangoapps/edraak_forus/models.py [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    ForUs users model
+    """
+    user = models.ForeignKey(User, unique=True, db_index=True)
+    date_created = models.DateTimeField(_('date forus profile created'), default=timezone.now)
+
+    def __unicode__(self):
+        return unicode(repr(self))
+
+    @staticmethod
+    # pylint: disable=missing-docstring
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/models.py]: Same thing, different class.

File: lms/djangoapps/edraak_forus/models.py [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    def create_for_user(user):
+        if user.is_staff or user.is_superuser:
+            log.warn('Cannot create profile for superusers and staff. email=`%s`', user.email)
+            raise Exception('Cannot create profile for superusers and staff')
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/models.py]: Can you create a separate exception class for this?

File: lms/djangoapps/edraak_forus/tests/tests.py [hunk 15]
```diff
@@ -0,0 +1,388 @@
+# pylint: disable=missing-docstring
+import json
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/tests/tests.py]: Try to group the imports.

 - Django
 - Native
 - edX
 - Edraak

What do you think?

File: lms/djangoapps/edraak_forus/tests/tests.py [hunk 16]
```diff
@@ -0,0 +1,388 @@
...
+    # pylint: disable=invalid-name
+    def _assertLoggedIn(self, msg_prefix=None):
+        res_dashboard = self.client.get(self.dashboard_url)
+        ## testing dashboard url to see if user is logged in
+        self.assertContains(res_dashboard, 'dashboard-main', msg_prefix=msg_prefix)
+
+    # pylint: disable=invalid-name
+    def _assertLoggedOut(self):
+        res_dashboard = self.client.get(self.dashboard_url)
+        self.assertEquals(res_dashboard.status_code, 302, 'User is not logged out.')
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/tests/tests.py]: Just use `self.assertRedirects()` looks like it has been invented later on :rofl:

File: lms/djangoapps/edraak_forus/tests/tests.py [hunk 17]
```diff
@@ -0,0 +1,388 @@
...
+        self._assertPathEquals(res_dashboard['Location'], '/login')
+
+    # pylint: disable=invalid-name
+    def _assertPathEquals(self, url_a, url_b):
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/tests/tests.py]: Perhaps remove this as well.

File: lms/djangoapps/edraak_forus/tests/tests.py [hunk 18]
```diff
@@ -0,0 +1,388 @@
...
+
+        self.client.logout()
+        self.client.session.clear()
+
+        self._assertLoggedOut()
+
+        res_auth_2 = self.client.get(self.auth_url, self._build_forus_params())
+
+        self.assertIsInstance(res_auth_2, HttpResponseRedirect)
+        self.assertTrue(
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/tests/tests.py]: Same thing for all redirects.

File: lms/djangoapps/edraak_forus/tests/tests.py [hunk 19]
```diff
@@ -0,0 +1,388 @@
...
+        res = self.client.get(self.url, {
+            'message': message,
+        })
+
+        self.assertNotContains(res, message, msg_prefix='The page is XSS vulnerable')
+        self.assertContains(res, escaped_message, msg_prefix='The page encodes the message incorrectly')
+
+
+@ddt.ddt
+@override_settings(LANGUAGE_CODE='en')
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/tests/tests.py]: That's bad, I think we're having a bug in the view/middleware in terms of language handling. I've never liked what I've done there, this has revealed a possible issue.

Let's study what edX does and try to do find alternative implementation.

File: lms/djangoapps/edraak_forus/tests/tests.py [hunk 20]
```diff
@@ -0,0 +1,388 @@
...
+    def test_missing_hmac(self):
+        params = build_forus_params(email=self.user_email, course_id=unicode(self.upcoming_course.id))
+        params.pop('forus_hmac', None)
+        # pylint: disable=invalid-name
+        with self.assertRaisesRegexp(ValidationError, 'The security check has failed on the provided parameters') as cm:
+            ValidateForusParams(params).validate()
+
+        self.assertIn('forus_hmac', cm.exception.message_dict)
+
+    def test_calculate_hmac(self):
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/tests/tests.py]: Not enough my friend. We need to have some hard data.

The code below merely duplicates the function logic, which is not good for testing. Fixtures is a better solution here.

Try to generate the fixtures using an external tool like bash. However the fixtures should be hardcoded, and the bash commands (for example) should be documented.

One complication would arise here, that you'd need to freeze the time, because the requests expires.

File: lms/djangoapps/edraak_forus/views.py [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+        user = User.objects.get(email=user_email)
+        user.is_active = True
+        user.save()
+
+        set_user_preference(user, LANGUAGE_KEY, lang)
+
+        ForusProfile.create_for_user(user)
+
+
+def message(request):
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/views.py]: I think it's time for this to be a class.

File: lms/templates/edraak_forus/auth.html [hunk 22]
```diff
@@ -0,0 +1,38 @@
+<%page expression_filter="h"/>
```
Comment (OmarIthawi) [inline lms/templates/edraak_forus/auth.html:2]: Did you manage to copy this from the latest version of `student_account/login_and_register.html`??

If not, please do.

Perhaps we can add a test to ensure the files are in sync and have the minimum amount of difference. Not sure about how to do this, but it looks like a good idea.

Anyway add a comment on this file for future maintainers.

File: lms/templates/edraak_forus/message.html [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+  </h1>
+  <p class="text-center">
+      ${Text(message)}
+  </p>
+
+  <hr />
+
+  <p class="text-center">
+    ${Text(_('Please go to {forus_link}, and try again. If the problem persists, please email us at {email}.')).format(
+        forus_link=HTML(u'<a href="{url}">{name}</a>').format(
```
Comment (OmarIthawi) [inline lms/templates/edraak_forus/message.html:21]: If you're using `HTML`, use `Text` for it's params. `Cmd+Shift+F` for examples.

File: lms/templates/edraak_forus/style.css [hunk 24]
```diff
@@ -0,0 +1,13 @@
+.form-field[class*=hidden] {
+    display: none;
+}
+
+.forus-notes {
+    margin-bottom: 20px;
+}
+
+.forus-notes p {
```
Comment (OmarIthawi) [inline lms/templates/edraak_forus/style.css]: Not very sscss.

File: lms/templates/edraak_forus/style.css [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+    font-size: .85em;
+    margin-top: 0;
+    margin-bottom: 10px;
+}
```
Comment (OmarIthawi) [inline lms/templates/edraak_forus/style.css]: 😄

File: lms/templates/edraak_forus/style.css [hunk 26]
```diff
@@ -0,0 +1,13 @@
+.form-field[class*=hidden] {
```
Comment (OmarIthawi) [inline lms/templates/edraak_forus/style.css]: CSS? I don't think it's a good idea. Just keep it inline then.

Or make it a first-class citizen SCSS file within the pipeline to ensure static file handling.

File: lms/urls.py [hunk 27]
```diff
@@ -158,6 +158,11 @@
         url(r'^sysadmin/', include('dashboard.sysadmin_urls')),
     )
 
+if settings.FEATURES["ENABLE_EDRAAK_FORUS"]:
```
Comment (OmarIthawi) [inline lms/urls.py]: Ooops!

**Review by OmarIthawi — CHANGES_REQUESTED** [review]
Case closed, insufficient evidences.

File: common/djangoapps/student/helpers.py (hunk 2 above)
Comment (Salomari1987) [reply common/djangoapps/student/helpers.py]: @OmarIthawi Actually if you look at the latest edX code, enrollment is still done this way, this is only related to LMS when a user chooses to enroll/unenroll in a course. I do not see how I can use the enrollment/api in this specific case.

File: lms/djangoapps/edraak_forus/views.py (hunk 21 above)
Comment (Salomari1987) [reply lms/djangoapps/edraak_forus/views.py]: العيال كبرت

Comment (OmarIthawi) [reply lms/djangoapps/edraak_forus/views.py]: طبعا طبعا

File: lms/djangoapps/edraak_forus/helpers.py (hunk 11 above)
Comment (Salomari1987) [reply lms/djangoapps/edraak_forus/helpers.py]: please elaborate

File: common/djangoapps/course_modes/helpers.py (hunk 1 above)
Comment (Salomari1987) [reply common/djangoapps/course_modes/helpers.py]: @OmarIthawi I think this is supposed to do the job? https://github.com/edx/course-discovery/pull/75/files

Comment (OmarIthawi) [reply common/djangoapps/course_modes/helpers.py]: @Salomari1987 I think so!

File: common/djangoapps/student/helpers.py (hunk 2 above)
Comment (OmarIthawi) [reply common/djangoapps/student/helpers.py]: Sure @Salomari1987. At the lms level. But we're already using the enrollment api in three places:

 - Marketing site
 - Android app
 - iOS app

So it should be safe to use it instead of this refactoring.

**Comment by OmarIthawi** [comment]
*This is completely optional and not a blocker to this PR.*

@Salomari1987 I remember that you've suggested this before, it looks like edX are already using the [`AccountCreationForm`](https://github.com/edx/edx-platform/search?q=AccountCreationForm&type=Code&utf8=%E2%9C%93) to validate user input. Great minds think alike 😉

**Comment by Salomari1987** [comment]
@OmarIthawi this is a possibility, and another would be to use an api to achieve the end result desired

File: common/djangoapps/course_modes/helpers.py (hunk 1 above)
Comment (OmarIthawi) [reply common/djangoapps/course_modes/helpers.py]: Let's remove it please. I've added a task for this:

 - [Integrate the Edraak marketing site with Eucalyptus](https://app.asana.com/0/256944342521670/303375741516935).

File: common/djangoapps/course_modes/helpers.py [hunk 28]
```diff
@@ -13,6 +15,43 @@
 DISPLAY_AUDIT = "audit"
 DISPLAY_PROFESSIONAL = "professional"
 
+COURSE_INFO_PAGE = "COURSE_INFO"
+SUCCESS_ENROLL_PAGE = "ENROLL_SUCCESS"
+
+
+def get_mktg_url(page):
```
Comment (OmarIthawi) [inline common/djangoapps/course_modes/helpers.py]: Please remove this as well, and use edX provided helpers, I've added an entry to the [Integrate the Edraak marketing site with Eucalyptus](https://app.asana.com/0/256944342521670/303375741516935) task on Asana to address this later.

File: common/djangoapps/course_modes/views.py [hunk 29]
```diff
@@ -104,7 +106,8 @@ def get(self, request, course_id, error=None):
         # in the "honor" track by this point, so we send the user
         # to the dashboard.
         if not CourseMode.has_verified_mode(modes):
-            return redirect(reverse('dashboard'))
+
+            return redirect(get_mktg_for_course(SUCCESS_ENROLL_PAGE, unicode(course_id)))
```
Comment (OmarIthawi) [inline common/djangoapps/course_modes/views.py]: You could use `reverse('dashboard')` here to address the comments above.

File: common/djangoapps/student/views.py [hunk 30]
```diff
@@ -145,7 +145,7 @@ def csrf_token(context):
     if token == 'NOTPROVIDED':
         return ''
     return (u'<div style="display:none"><input type="hidden"'
-            ' name="csrfmiddlewaretoken" value="%s" /></div>' % (token))
+            ' name="csrfmiddlewaretoken" value="%s" /></div>' % token)
```
Comment (OmarIthawi) [inline common/djangoapps/student/views.py]: Please undo this change. The file is unrelated and we shouldn't introduce noise.

File: common/test/acceptance/tests/studio/test_studio_components.py [hunk 31]
```diff
@@ -1,6 +1,9 @@
 """
 Acceptance tests for adding components in Studio.
 """
+
+from flaky import flaky
```
Comment (OmarIthawi) [inline common/test/acceptance/tests/studio/test_studio_components.py]: Please rebase on master, this is not specific to this branch, right?

File: lms/djangoapps/edraak_forus/admin.py [hunk 32]
```diff
@@ -0,0 +1,54 @@
...
+    readonly_fields = ('user', 'date_created')
+    search_fields = ('user__email', 'user__username')
+
+    def edraak_user(self, profile):
+        """
+        Return link to user
+        """
+        user = profile.user
+        return HTML(u'<a href="{url}">{name}</a>').format(
+            name=escape(user.username),
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/admin.py]: Please use `Text` instead of `escape`.

File: lms/djangoapps/edraak_forus/helpers.py [hunk 33]
```diff
@@ -0,0 +1,318 @@
...
+        self.validate_user_email()
+        self.validate_user_profile()
+        self.validate_course()
+        self.validate_request_time()
+
+        if len(self.errors):
+            raise ValidationError(self.errors)
+
+
+def build_message(params):
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/helpers.py]: Very weird naming! Did I name it so?

File: lms/djangoapps/edraak_forus/models.py [hunk 34]
```diff
@@ -0,0 +1,55 @@
...
+import logging
+from django.db import models
+from django.contrib.auth.models import User
+from django.utils.translation import ugettext_lazy as _
+from django.utils import timezone
+
+log = logging.getLogger(__name__)
+
+
+class ForusCreateUserProfile(Exception):
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/models.py]: This is super unconventional. Use something like:

 - `ForusError`

File: lms/djangoapps/edraak_forus/models.py [hunk 35]
```diff
@@ -0,0 +1,55 @@
...
+        return unicode(repr(self))
+
+    @staticmethod
+    def create_for_user(user):
+        """
+        Create ForUs users
+        """
+        if user.is_staff or user.is_superuser:
+            log.warn('Cannot create profile for superusers and staff. email=`%s`', user.email)
+            raise ForusCreateUserProfile
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/models.py]: Always add an exception message.

```
ForusCreateUserProfile('...')
```

File: lms/djangoapps/edraak_forus/models.py [hunk 36]
```diff
@@ -0,0 +1,55 @@
...
+    """
+    ForUs users model
+    """
+    user = models.ForeignKey(User, unique=True, db_index=True)
+    date_created = models.DateTimeField(_('date forus profile created'), default=timezone.now)
+
+    def __unicode__(self):
+        return unicode(repr(self))
+
+    @staticmethod
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/models.py]: This can be a `@classmethod` instead.

File: lms/djangoapps/edraak_forus/models.py [hunk 37]
```diff
@@ -0,0 +1,55 @@
...
+        Create ForUs users
+        """
+        if user.is_staff or user.is_superuser:
+            log.warn('Cannot create profile for superusers and staff. email=`%s`', user.email)
+            raise ForusCreateUserProfile
+
+        forus_profile = ForusProfile(user=user)
+        forus_profile.save()
+
+    @staticmethod
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/models.py]: This can be a `@classmethod` instead.

File: lms/djangoapps/edraak_forus/tests/helpers.py [hunk 38]
```diff
@@ -0,0 +1,29 @@
+"""
+Edraak ForUs helpers
+"""
+from datetime import datetime
+
+TIME_FORMAT = '%Y-%m-%dT%H:%M:%S'
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/tests/helpers.py]: This can be imported from the helpers file, `edraak_forus/helpers.py`, no?

File: lms/djangoapps/edraak_forus/tests/tests.py [hunk 39]
```diff
@@ -0,0 +1,508 @@
...
+
+    @ddt.data('', ' ', 'hello')
+    def test_invalid_level_of_education(self, bad_value):
+        self.assert_validated_data(
+            field_id='level_of_education',
+            bad_value=bad_value,
+            exception_regexp='.*Invalid.*level of education.*',
+        )
+
+    @ddt.data('', ' ', 'x', 'hello')
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/tests/tests.py:384]: Please add a task for the gender validation in the Open edX Asana project.

File: lms/djangoapps/edraak_forus/tests/tests.py [hunk 40]
```diff
@@ -0,0 +1,508 @@
...
+        This tests the calculate hmac helper in ForUs,
+        The time is forzen to '2017-01-14 03:21:34'
+        Course_id is frozen to org.51/course_51/Run_51
+
+        Use bash tool to generate expected hmac:
+            1. store msg to be hashed: ```export \
+                msg="course_id=org.51/course_51/Run_51;email=forus.user.faramvalidatortest@example.com; \
+                name=Abdulrahman (ForUs);enrollment_action=enroll; \
+                country=JO;level_of_education=hs;gender=m;year_of_birth=1989;lang=en;time=2017-01-14T03:21:34"```
+            2. hash msg: ```echo -n $msg | openssl dgst -sha256 -hmac "dummy_auth_secret_key"```
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/tests/tests.py]: Great work here!

File: lms/envs/aws.py [hunk 41]
```diff
@@ -864,3 +864,12 @@
 APP_UPGRADE_CACHE_TIMEOUT = ENV_TOKENS.get('APP_UPGRADE_CACHE_TIMEOUT', APP_UPGRADE_CACHE_TIMEOUT)
 
 AFFILIATE_COOKIE_NAME = ENV_TOKENS.get('AFFILIATE_COOKIE_NAME', AFFILIATE_COOKIE_NAME)
+
+################################ Edraak Apps ################################
+
+if FEATURES.get('ENABLE_EDRAAK_FORUS'):
```
Comment (OmarIthawi) [inline lms/envs/aws.py]: Please rebase on `master` and move all the `/envs/` (the four of them) to the other `Edraak Apps` sections.

File: lms/templates/edraak_forus/auth.html (hunk 22 above)
Comment (OmarIthawi) [reply lms/templates/edraak_forus/auth.html:2]: Still not sure that this has been migrated. We agreed to do one of the following approaches:

 - Add some kind of tests to ensure only needed modifications are done.
 - Refactor into three files: a) Base, b) edX auth, c) ForUs auth.

File: lms/templates/edraak_forus/auth.html [hunk 42]
```diff
@@ -0,0 +1,63 @@
...
+    </script>
+    % for template_name in ["account", "access", "form_field", "login", "institution_login", "institution_register", "password_reset", "hinted_login"]:
+        <script type="text/template" id="${template_name}-tpl">
+            <%static:include path="student_account/${template_name}.underscore" />
+        </script>
+% endfor
+</%block>
+
+<style>
+    /* TODO: Move this into a CSS file */
```
Comment (OmarIthawi) [inline lms/templates/edraak_forus/auth.html]: Should we move this now? Or keep it here?

**Comment by OmarIthawi** [comment]
@Salomari1987 I will also test it locally and let you know about the results.

File: lms/envs/test.py [hunk 43]
```diff
@@ -525,6 +525,18 @@
 
 NOTES_DISABLED_TABS = []
 
+########################## EDRAAK FORUS ###############################
+
+# Enable Edraak ForUs app for tests
+FEATURES['ENABLE_EDRAAK_FORUS'] = True
```
Comment (OmarIthawi) [inline lms/envs/test.py]: Let's keep the Edraak convention and name it like the following:

 - `FEATURES['EDRAAK_FORUS_APP']`.

Everywhere please 😃

File: lms/djangoapps/edraak_forus/views.py [hunk 44]
```diff
@@ -0,0 +1,291 @@
...
+            'login_redirect_url': redirect_to,  # This gets added to the query string of the "Sign In" button in header
+            'responsive': True,
+            'allow_iframing': True,
+            'disable_courseware_js': True,
+            'disable_footer': True,
+        }
+
+        return render_to_response('edraak_forus/auth.html', context)
+
+    def dispatch(self, request, *args, **kwargs):
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/views.py]: @Salomari1987 I've removed this function and the tests are still passing. Let's just remove this and make the `lang` param practically effectless.

Comment (OmarIthawi) [reply lms/djangoapps/edraak_forus/views.py]: @Salomari1987 I think we can do this later:

 - [Activate forus API lang param with integration tests with edraak_i18n app](https://app.asana.com/0/256944342521670/305314536764578).

File: lms/djangoapps/edraak_forus/models.py [hunk 45]
```diff
@@ -0,0 +1,55 @@
...
+    Cannot create profile for superusers and staff
+    """
+    pass
+
+
+class ForusProfile(models.Model):
+    """
+    ForUs users model
+    """
+    user = models.ForeignKey(User, unique=True, db_index=True)
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/models.py]: I know it's my fault! but this should be a `OneToOneField` according to Django:

```
System check identified some issues:

WARNINGS:
edraak_forus.ForusProfile.user: (fields.W342) Setting unique=True on a ForeignKey has the same effect as using a OneToOneField.
	HINT: ForeignKey(unique=True) is usually better served by a OneToOneField.
```

Please delete the migrations before making a new one.

**Comment by OmarIthawi** [comment]
False alarm. The registration works fine, I've added a [step in the checklist](https://github.com/Edraak/edraak-platform/wiki/Checklist-to-Get-a-Working-Edraak-Platform#making-kafdforus-works) to make sure it works fine.

----

@Salomari1987 I've got a bug for you.

## Bug no. 1

**STRs:**
 - Use the [testing tool](https://github.com/Edraak/kafd-forus-tools)
 - Run a happy scenario, just try to login using ForUs


**Expected:**
 - To login

**Actual:**
 - It redirect to the login page and doesn't login!
 - The user is not created in the database
 - The ForUs profile is missing (of course)
 - No error message or anything for the user

File: lms/djangoapps/edraak_forus/tests/tests.py (hunk 39 above)
Comment (Salomari1987) [reply lms/djangoapps/edraak_forus/tests/tests.py:384]: FYI [Add gender validation for alternative genders](https://app.asana.com/0/256944342521670/310071418198111)

**Comment by OmarIthawi** [comment]
@Salomari1987 I turns out that this needs the `country` field enabled, so it's not a bug. Please address the remaining issues so we can merge this one.

**Comment by OmarIthawi** [comment]
As you've suggestion, you'll add two bok choy tests:

 - Happy scenario when a new user successfully enrolls
 - (Optional) Happy scenario when a returning user successfully enrolls
 - Ensure that error messages are handled properly through the messages view as opposed to silently redirect to the login page

**Comment by Salomari1987** [comment]
@OmarIthawi please review it man

File: common/djangoapps/student/views.py (hunk 30 above)
Comment (OmarIthawi) [reply common/djangoapps/student/views.py]: @Salomari1987 ping.

File: common/test/acceptance/tests/studio/test_studio_components.py (hunk 31 above)
Comment (OmarIthawi) [reply common/test/acceptance/tests/studio/test_studio_components.py]: Ping @Salomari1987.

**Comment by OmarIthawi** [comment]
This is great @Salomari1987. Seriously! I'm really happy with what this module has become.

I hope that we will have other chances to do more interesting features in the future.

File: lms/djangoapps/edraak_forus/helpers.py [hunk 46]
```diff
@@ -0,0 +1,378 @@
...
+        'year_of_birth',
+    )
+
+    _hidden_fields = (
+        'goals',  # Hide goals because the ForUs people didn't like to
+        # ask their students about their goals in Edraak.
+        'password',
+    )
+
+    def __init__(self):
```
Comment (OmarIthawi) [inline lms/djangoapps/edraak_forus/helpers.py]: Could you please remove this init? It's useless.

**Comment by OmarIthawi** [comment]
@Salomari1987 three minor niptucks and we're good to go!

**Comment by Salomari1987** [comment]
@OmarIthawi ready for review

**Review by OmarIthawi — CHANGES_REQUESTED** [review]
@Salomari1987 there's a duplicate `from flaky import flaky` in the `test_studio_components.py` file. Please remove the redundant one.

**Review by OmarIthawi — APPROVED** [review]

**Comment by OmarIthawi** [comment]
@Salomari1987 just rebased it and resolved the conflicts. Hope the tests passes!

**Comment by OmarIthawi** [comment]
🎉 And that's your birthday gift @Salomari1987 🎂
